File: mobile/lib/domain/models/money.dart
import 'package:intl/intl.dart';

/// An amount of money, as minor units and a currency.
///
/// The single money type in the customer app, and the only place a price is
/// turned into something a person reads. Three rules are enforced here so that
/// no screen has to remember them:
///
/// 1. **Never a double.** ₹249.90 stored as a binary float is 249.90000000000001,
///    and a menu of them summed at checkout is off by a rupee for reasons no
///    customer will accept. The wire format is an integer count of paise and
///    the app keeps it that way.
///
/// 2. **The currency travels with the amount.** An integer on its own is not a
///    price; 24900 is ₹249 or $249 depending on something a widget three files
///    away cannot see.
///
/// 3. **The symbol is not hard-coded into logic.** [format] is the only thing
///    in the app that decides what a price looks like, and it asks `intl` using
///    the customer's own locale rather than concatenating a rupee sign.
class Money implements Comparable<Money> {
  const Money({required this.amountMinor, required this.currency});

  /// The amount in the currency's smallest unit — paise for INR.
  final int amountMinor;

  /// ISO 4217, upper case.
  final String currency;

  /// How many minor units make one major unit.
  ///
  /// Not every currency is a hundred: the yen has no minor unit at all, and
  /// treating ¥500 as five yen would be a hundredfold error. Unknown currencies
  /// fall back to two places, which is right far more often than it is wrong.
  static const Map<String, int> _subunits = <String, int>{
    'INR': 100,
    'USD': 100,
    'EUR': 100,
    'GBP': 100,
    'JPY': 1,
  };

  int get subunits => _subunits[currency] ?? 100;

  /// The number of decimal places this currency is written with.
  int get decimalDigits => switch (subunits) {
    1 => 0,
    1000 => 3,
    _ => 2,
  };

  bool get isZero => amountMinor == 0;

  /// True when the amount has no fractional part — ₹249 rather than ₹249.50.
  bool get isWhole => amountMinor % subunits == 0;

  /// Reads `{"amount_minor": 24900, "currency": "INR"}`.
  ///
  /// Returns null rather than guessing. A price that cannot be read is not
  /// rendered as zero and not rendered as "—" next to an "Add" button: the
  /// caller drops the item, because an item whose price is unknown is an item
  /// nobody can be asked to buy.
  static Money? fromJson(Object? json) {
    if (json is! Map<String, dynamic>) return null;

    final Object? raw = json['amount_minor'];
    final Object? currency = json['currency'];

    // Deliberately not `num`: a double here means the server has started
    // sending rupees, and silently rounding it would introduce exactly the
    // error this class exists to prevent.
    final int? amount = switch (raw) {
      final int value => value,
      final String value => int.tryParse(value),
      _ => null,
    };

    if (amount == null || amount < 0) return null;
    if (currency is! String || currency.length != 3) return null;

    return Money(amountMinor: amount, currency: currency.toUpperCase());
  }

  /// The price as a person reads it, in their locale.
  ///
  /// Whole amounts lose their decimals — a menu of "₹249.00" reads like a
  /// spreadsheet — but anything with paise keeps them, because ₹249.50 shown
  /// as ₹249 is a lie about the price.
  String format({String? locale}) {
    final NumberFormat formatter = NumberFormat.simpleCurrency(
      locale: locale ?? Intl.getCurrentLocale(),
      name: currency,
      decimalDigits: isWhole ? 0 : decimalDigits,
    );

    return formatter.format(amountMinor / subunits);
  }

  /// What a screen reader should say.
  ///
  /// "₹249" is announced by some engines as "rupee two four nine". Spelling the
  /// currency out is longer to read and correct to hear.
  String spokenLabel({String? locale}) {
    final String amount = NumberFormat.decimalPattern(
      locale ?? Intl.getCurrentLocale(),
    ).format(isWhole ? amountMinor ~/ subunits : amountMinor / subunits);

    return switch (currency) {
      'INR' => '$amount rupees',
      'USD' => '$amount dollars',
      'EUR' => '$amount euros',
      'GBP' => '$amount pounds',
      'JPY' => '$amount yen',
      _ => '$amount $currency',
    };
  }

  @override
  int compareTo(Money other) {
    // Comparing two currencies by their raw integers would rank ¥500 above
    // ₹400 for reasons that have nothing to do with value. Refused rather than
    // answered wrongly.
    if (other.currency != currency) {
      throw ArgumentError('cannot compare $currency with ${other.currency}');
    }

    return amountMinor.compareTo(other.amountMinor);
  }

  @override
  bool operator ==(Object other) =>
      other is Money &&
      other.amountMinor == amountMinor &&
      other.currency == currency;

  @override
  int get hashCode => Object.hash(amountMinor, currency);

  @override
  String toString() => 'Money($amountMinor $currency)';
}

File: backend/app/Support/Money.php
<?php

declare(strict_types=1);

namespace App\Support;

use InvalidArgumentException;

/**
 * An amount of money.
 *
 * Integer minor units — paise for rupees, cents for dollars — and a currency,
 * always together. Two rules, and both of them are the reason this class exists
 * rather than a pair of loose variables.
 *
 * **Never a float.** A binary floating-point number cannot represent 0.10
 * exactly. A menu price that has been through one is a bill that does not add
 * up, and the error compounds the moment a later module starts summing a cart.
 *
 * **Never an amount without its currency.** "24900" is not a price. Storing the
 * two apart is how a rupee menu ends up rendered in dollars the first time this
 * platform crosses a border.
 *
 * Deliberately *not* a formatter. What a price looks like — the symbol, the
 * grouping, whether the paise are shown — is a locale decision, and the server
 * does not know the customer's. The API sends the amount and the currency; the
 * client renders them. See `25-customer-menu-browsing.md`.
 */
final readonly class Money
{
    /** Currencies this platform knows how to divide. */
    private const SUBUNITS = [
        'INR' => 100,
        'USD' => 100,
        'EUR' => 100,
        'GBP' => 100,
        // Zero-decimal currencies exist and are not a rounding edge case: 100
        // yen is 100, not 10 000. Listed so the day one is needed the answer is
        // here rather than assumed.
        'JPY' => 1,
    ];

    private function __construct(
        public int $minor,
        public string $currency,
    ) {}

    /**
     * @throws InvalidArgumentException
     */
    public static function fromMinor(int $minor, string $currency = 'INR'): self
    {
        $code = strtoupper(trim($currency));

        if (! isset(self::SUBUNITS[$code])) {
            throw new InvalidArgumentException("Unknown currency: {$currency}");
        }

        if ($minor < 0) {
            // A negative menu price is not a discount, it is a corrupt row. The
            // column is unsigned so one cannot be stored; this catches the case
            // where a value reaches here from somewhere that is not the column.
            throw new InvalidArgumentException('An amount cannot be negative.');
        }

        return new self($minor, $code);
    }

    /**
     * Reads a stored price, or null if the row cannot be trusted.
     *
     * Used on the customer path, where a single corrupt row must not take the
     * whole menu down. The caller omits the item rather than rendering a price
     * nobody can defend.
     */
    public static function tryFromMinor(mixed $minor, mixed $currency): ?self
    {
        if (! is_int($minor) && ! (is_string($minor) && ctype_digit($minor))) {
            return null;
        }

        if (! is_string($currency)) {
            return null;
        }

        try {
            return self::fromMinor((int) $minor, $currency);
        } catch (InvalidArgumentException) {
            return null;
        }
    }

    /** How many minor units make one major unit of this currency. */
    public function subunits(): int
    {
        return self::SUBUNITS[$this->currency];
    }

    /** True when the amount is a whole number of rupees, dollars, and so on. */
    public function isWhole(): bool
    {
        return $this->minor % $this->subunits() === 0;
    }

    public function isZero(): bool
    {
        return $this->minor === 0;
    }

    /**
     * @return array{amount_minor: int, currency: string}
     */
    public function toApiArray(): array
    {
        return [
            'amount_minor' => $this->minor,
            'currency' => $this->currency,
        ];
    }
}
